docs/how_to/pipeline_python.md

```diff
@@ -0,0 +1,151 @@
+
+## Declaratively specifying a pipeline
+
+The declarative specification approach looks similar to a YAML file hierarchy, but consists of nested Python dictionary and list objects.
+
+```{code-block} python
+import lumen
+from lumen.pipeline import Pipeline
+
+data_url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-07-28/penguins.csv'
+
+pipeline = Pipeline.from_spec({
+    'source': {
+        'type': 'file',
+        'tables': {
+            'penguins': data_url
+        }
+    },
+    'filters': [
+        {'type': 'widget', 'field': 'species'},
+        {'type': 'widget', 'field': 'island'},
+        {'type': 'widget', 'field': 'sex'},
+        {'type': 'widget', 'field': 'year'}
+    ],
+    'transforms': [
+        {'type': 'aggregate', 'method': 'mean', 'by': ['species', 'sex', 'year']}
+    ]
+})
+```
+
+## Programmatically specifying a pipeline
+
+The programmatic specification approach uses Lumen objects to build the pipeline step by step.
+
+### Add source
+First, add a valid `Source` to your `Pipeline`. A common choice is `FileSource`, which can load CSV, Excel, JSON and Parquet files, but see the [Source Reference](../architecture//source.html#:~:text=Source%20queries%20data.-,Source%20types%23,-class%20lumen.sources) for all options.
+
+```{code-block} python
+from lumen.pipeline import Pipeline
+from lumen.sources import FileSource
+
+data_url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-07-28/penguins.csv'
+
+pipeline = Pipeline(source=FileSource(tables={'penguins': data_url}), table='penguins')
+```
+
+:::{admonition} Preview the data
+:class: note
+At any point after defining the source in your pipeline, you can inspect the data in a notebook with `pipeline.data`
+:::
+:::{dropdown} `pipeline.data`
+:animate: fade-in-slide-down
+![data preview](../_static/pipeline_data.png)
+:::
+
+### Add filter
+Next, you can add `widgets` for certain columns of your source. When displaying the dashboard, these widgets will allows your dashboard users to `filter` the data. See the [Filter Reference](../architecture/filter) for all options.
+
+```{code-block} python
+:emphasize-lines: 9-12
+from lumen.pipeline import Pipeline
+from lumen.sources import FileSource
+
+data_url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-07-28/penguins.csv'
+
+pipeline = Pipeline(source=FileSource(tables={'penguins': data_url}), table='penguins')
+
+# Filters
+pipeline.add_filter('widget', field='species')
+pipeline.add_filter('widget', field='island')
+pipeline.add_filter('widget', field='sex')
+pipeline.add_filter('widget', field='year')
+```
+
+### Add transform
+Now you can apply a `transform` to the data, such as computing the mean or selecting certain columns. See the [Transform Reference](../architecture/transform) for more.
+
+```{code-block} python
+:emphasize-lines: 14-15
```

Comment:
Will have to see if I can get emphasis working in the pyodide extension so we can have live output here too.